torchtune/utils/memory.py

```diff
@@ -240,3 +241,79 @@ def log_memory_stats(stats: Dict[str, float]) -> None:
         f"\n\tGPU peak memory reserved: {stats['peak_memory_reserved']:.2f} GiB"
         f"\n\tGPU peak memory active: {stats['peak_memory_active']:.2f} GiB"
     )
+
+
+# named tuple for passing GPU memory stats for logging
+GPUMemStats = namedtuple(
+    "GPUMemStats",
+    [
+        "max_active_gib",
+        "max_active_pct",
+        "max_reserved_gib",
+        "max_reserved_pct",
+        "num_alloc_retries",
+        "num_ooms",
+    ],
+)
+
+
+class GPUMemoryMonitor:
+    def __init__(self, device: str = "cuda:0"):
+        self.device = torch.device(device)  # device object
+        self.device_name = torch.cuda.get_device_name(self.device)
+        self.device_index = torch.cuda.current_device()
+        self.device_capacity = torch.cuda.get_device_properties(
+            self.device
+        ).total_memory
+        self.device_capacity_gib = self._to_gib(self.device_capacity)
+
+        torch.cuda.reset_peak_memory_stats()
+        torch.cuda.empty_cache()
+
+    def _to_gib(self, memory_in_bytes):
+        # NOTE: GiB (gibibyte) is 1024, vs GB is 1000
+        _gib_in_bytes = 1024 * 1024 * 1024
+        memory_in_gib = memory_in_bytes / _gib_in_bytes
+        return memory_in_gib
+
+    def _to_pct(self, memory):
+        return 100 * memory / self.device_capacity
+
+    def get_peak_stats(self):
+        cuda_info = torch.cuda.memory_stats(self.device)
+
+        max_active = cuda_info["active_bytes.all.peak"]
+        max_active_gib = self._to_gib(max_active)
+        max_active_pct = self._to_pct(max_active)
+
+        max_reserved = cuda_info["reserved_bytes.all.peak"]
+        max_reserved_gib = self._to_gib(max_reserved)
+        max_reserved_pct = self._to_pct(max_reserved)
+
+        num_retries = cuda_info["num_alloc_retries"]
+        num_ooms = cuda_info["num_ooms"]
```

Comment:
Hi @felipemello1  - 
1 - how can num_retries be used:  The num retries means the number of times cudaCachingAllocator did not have enough memory to dispense for a request.  Therefore, it drops all memory and then re-requests it via cudaMalloc to see if the lack of available memory is due to fragmentation...often it is, so training can continue but it is a time hit due to drop and restart all caching. 
Generally, even a single retry is a perf hit, but certainly only seeing 1 vs 5 in an iter is the degree of perf hit.  Thus, it just provides more data as to how badly/often you are taking perf hits.

Reply:
2 - can num_oom be >0 and the run not fail?  Yes, usually not in standard training but apparently in notebooks OOMs can happen and continue.  
This metric is generally not that useful since we are inherently running in non-interactive training/tuning, so that's why it's not displayed atm.  We can just remove this from tracking though.